File: js/task3/S3_PIPE_01.js
// S3_PIPE_01 — pipe()

function pipe(...fns) {
  return (input) => {
    let value = input;
    for (const fn of fns) {
      value = fn(value);
    }
    return value;
  };
}

// Tests
const add1 = (n) => n + 1;
const double = (n) => n * 2;
const toStr = (n) => String(n);

const f = pipe(add1, double, toStr);

console.log(f(3)); // "8"  -> (3+1)=4, *2=8, to string

File: python/task3c/S3_PIPE_01.py
# S3_PIPE_01 — Implement pipe(*fns)

def pipe(*fns):
    def runner(x):
        value = x
        for fn in fns:
            value = fn(value)
        return value
    return runner


def main():
    add2 = lambda n: n + 2
    times3 = lambda n: n * 3
    minus5 = lambda n: n - 5

    p = pipe(add2, times3, minus5)

    # ((10 + 2) * 3) - 5 = 31
    print("pipe result:", p(10))


if __name__ == "__main__":
    main()
